Flaky-test: BrokerClientIntegrationTest.testConsumerWithPooledMessagesWithReader

<!--- 

Instructions for reporting a flaky test using this issue template:

1. Replace [test class] in title and body with the test class name
2. Replace [test method] in title and body with the test method that failed. Multiple methods are flaky, remove the content that refers to the test method.
3. Replace "url here" with a url to an example failure. In the Github Actions workflow run logs, you can right click on the line number to copy a link to the line. Example of such url is https://github.com/apache/pulsar/pull/8892/checks?check_run_id=1531075794#step:9:377 . The logs are available for a limited amount of time (usually for a few weeks).
4. Replace "relevant parts of the exception stacktrace here" with the a few lines of the stack trace that shows at leat the exception message and the line of test code where the stacktrace occured.
5. Replace "full exception stacktrace here" with the full exception stacktrace from logs. This section will be hidded by default.
6. Remove all unused fields / content to unclutter the reported issue. Remove this comment too.

-->
BrokerClientIntegrationTest is flaky. The testConsumerWithPooledMessagesWithReader test method fails sporadically.

[example failure](https://github.com/apache/pulsar/pull/11748/checks?check_run_id=3399716676#step:9:2325)

```
Error:  testConsumerWithPooledMessagesWithReader(org.apache.pulsar.client.impl.BrokerClientIntegrationTest)  Time elapsed: 2.234 s  <<< FAILURE!
java.lang.AssertionError: expected [true] but found [false]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertTrue(Assert.java:45)
	at org.testng.Assert.assertTrue(Assert.java:55)
	at org.apache.pulsar.client.impl.BrokerClientIntegrationTest.testConsumerWithPooledMessagesWithReader(BrokerClientIntegrationTest.java:1004)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

--- a/pom.xml
+++ b/pom.xml
@@ -1312,7 +1312,7 @@ flexible messaging model and an intuitive client API.</description>
         <artifactId>maven-surefire-plugin</artifactId>
         <configuration>
           <argLine> -Xmx1G -XX:+UseG1GC
-            -Dpulsar.allocator.pooled=false
+            -Dpulsar.allocator.pooled=true
             -Dpulsar.allocator.leak_detection=Advanced
             -Dpulsar.allocator.exit_on_oom=false
             -Dio.netty.tryReflectionSetAccessible=true

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java
@@ -105,7 +105,7 @@ public class TransactionMarkerDeleteTest extends BrokerTestBase{
                 .setProducerName("test")
                 .setSequenceId(0);
 
-        ByteBuf payload = PooledByteBufAllocator.DEFAULT.buffer(0);
+        ByteBuf payload = PooledByteBufAllocator.DEFAULT.heapBuffer(0);
 
         payload = Commands.serializeMetadataAndPayload(Commands.ChecksumType.Crc32c,
                     msgMetadata, payload);

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/TransactionMarkerDeleteTest.java
@@ -26,12 +26,10 @@ import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertTrue;
-import java.util.Collections;
-import java.util.concurrent.TimeUnit;
-
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.PooledByteBufAllocator;
-import io.netty.buffer.Unpooled;
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
 import org.apache.bookkeeper.mledger.ManagedCursor;
 import org.apache.bookkeeper.mledger.ManagedLedger;
 import org.apache.bookkeeper.mledger.Position;
@@ -55,7 +53,7 @@ import org.testng.annotations.Test;
 import org.testng.collections.Sets;
 
 @Test(groups = "broker")
-public class TransactionMarkerDeleteTest extends BrokerTestBase{
+public class TransactionMarkerDeleteTest extends BrokerTestBase {
 
     @BeforeMethod
     @Override
@@ -96,7 +94,6 @@ public class TransactionMarkerDeleteTest extends BrokerTestBase{
     }
 
 
-
     @Test
     public void testMarkerDelete() throws Exception {
 
@@ -105,10 +102,10 @@ public class TransactionMarkerDeleteTest extends BrokerTestBase{
                 .setProducerName("test")
                 .setSequenceId(0);
 
-        ByteBuf payload = PooledByteBufAllocator.DEFAULT.heapBuffer(0);
+        ByteBuf payload = PooledByteBufAllocator.DEFAULT.buffer(0);
 
         payload = Commands.serializeMetadataAndPayload(Commands.ChecksumType.Crc32c,
-                    msgMetadata, payload);
+                msgMetadata, payload);
 
         ManagedLedger managedLedger = pulsar.getManagedLedgerFactory().open("test");
         PersistentTopic topic = mock(PersistentTopic.class);
@@ -119,15 +116,16 @@ public class TransactionMarkerDeleteTest extends BrokerTestBase{
         PersistentSubscription persistentSubscription = new PersistentSubscription(topic, "test",
                 managedLedger.openCursor("test"), false);
 
-        Position position1 = managedLedger.addEntry(payload.array());
-        Position markerPosition1 = managedLedger.addEntry(Markers
-                .newTxnCommitMarker(1, 1, 1).array());
+        byte[] payloadBytes = toBytes(payload);
+        Position position1 = managedLedger.addEntry(payloadBytes);
+        Position markerPosition1 = managedLedger.addEntry(toBytes(Markers
+                .newTxnCommitMarker(1, 1, 1)));
 
-        Position position2 = managedLedger.addEntry(payload.array());
-        Position markerPosition2 = managedLedger.addEntry(Markers
-                .newTxnAbortMarker(1, 1, 1).array());
+        Position position2 = managedLedger.addEntry(payloadBytes);
+        Position markerPosition2 = managedLedger.addEntry(toBytes(Markers
+                .newTxnAbortMarker(1, 1, 1)));
 
-        Position position3 = managedLedger.addEntry(payload.array());
+        Position position3 = managedLedger.addEntry(payloadBytes);
 
         assertEquals(cursor.getNumberOfEntriesInBacklog(true), 5);
         assertTrue(((PositionImpl) cursor.getMarkDeletedPosition()).compareTo((PositionImpl) position1) < 0);
@@ -150,14 +148,14 @@ public class TransactionMarkerDeleteTest extends BrokerTestBase{
                         .compareTo((PositionImpl) markerPosition2) == 0);
 
         // add consequent marker
-        managedLedger.addEntry(Markers
-                .newTxnCommitMarker(1, 1, 1).array());
+        managedLedger.addEntry(toBytes(Markers
+                .newTxnCommitMarker(1, 1, 1)));
 
-        managedLedger.addEntry(Markers
-                .newTxnAbortMarker(1, 1, 1).array());
+        managedLedger.addEntry(toBytes(Markers
+                .newTxnAbortMarker(1, 1, 1)));
 
-        Position markerPosition3 = managedLedger.addEntry(Markers
-                .newTxnAbortMarker(1, 1, 1).array());
+        Position markerPosition3 = managedLedger.addEntry(toBytes(Markers
+                .newTxnAbortMarker(1, 1, 1)));
 
         // ack with transaction, then commit this transaction
         persistentSubscription.transactionIndividualAcknowledge(new TxnID(0, 0),
@@ -170,5 +168,12 @@ public class TransactionMarkerDeleteTest extends BrokerTestBase{
                 ((PositionImpl) persistentSubscription.getCursor().getMarkDeletedPosition())
                         .compareTo((PositionImpl) markerPosition3) == 0);
 
+    }
+
+    static byte[] toBytes(ByteBuf byteBuf) {
+        byte[] buf = new byte[byteBuf.readableBytes()];
+        byteBuf.readBytes(buf);
+        byteBuf.release();
+        return buf;
     }
 }

